src/field_theory/galois.lean

```diff
@@ -0,0 +1,233 @@
+
+## Main results
+
+- `fixing_subgroup_of_fixed_field` : If `E/F` is finite dimensional (but not necessarily Galois)
+  then `fixing_subgroup (fixed_field H) = H`
+- `fixed_field_of_fixing_subgroup`: If `E/F` is finite dimensional and Galois
+  then `fixed_field (fixing_subgroup K) = K`
+Together, these two result prove the Galois correspondence
+-/
+
+noncomputable theory
+open_locale classical
+
+open finite_dimensional alg_equiv
+
+namespace galois
+
+section
+
+variables (F : Type*) [field F] (E : Type*) [field E] [algebra F E]
+
+/-- A field extension E/F is galois if it is both separable and normal -/
+@[class] def is_galois : Prop := is_separable F E ∧ normal F E
+
+instance of_fixed_field (G : Type*) [group G] [fintype G] [mul_semiring_action G E] :
+  is_galois (mul_action.fixed_points G E) E :=
+⟨fixed_points.separable G E, fixed_points.normal G E⟩
+
+instance aut : group (E ≃ₐ[F] E) :=
+{ mul := λ ϕ ψ, ψ.trans ϕ,
+  mul_assoc := λ ϕ ψ χ, rfl,
+  one := 1,
+  one_mul := λ ϕ, by {ext, refl},
+  mul_one := λ ϕ, by {ext, refl},
+  inv := symm,
+  mul_left_inv := λ ϕ, by {ext, exact symm_apply_apply ϕ a} }
+
+lemma is_galois_implies_card_aut_eq_findim [finite_dimensional F E] [h : is_galois F E] :
+  fintype.card (E ≃ₐ[F] E) = findim F E :=
+begin
+  cases field.exists_primitive_element h.1 with α hα,
+  cases h.1 α with H1 h_separable,
+  cases h.2 α with H2 h_splits,
+  have switch : (⊤ : intermediate_field F E).to_subalgebra.to_submodule = ⊤ :=
+    by { ext, exact iff_of_true intermediate_field.mem_top submodule.mem_top },
+  rw [←findim_top, ←switch],
+  change fintype.card (E ≃ₐ[F] E) = findim F (⊤ : intermediate_field F E),
+  replace h_splits : polynomial.splits (algebra_map F F⟮α⟯) (minimal_polynomial H2),
+  { rw hα,
+    let map : E →+* (⊤ : intermediate_field F E) :=
+    { to_fun := λ x, ⟨x, intermediate_field.mem_top⟩,
+      map_one' := rfl,
+      map_mul' := λ _ _, rfl,
+      map_zero' := rfl,
+      map_add' := λ _ _, rfl },
+    rw (show algebra_map F (⊤ : intermediate_field F E) = map.comp (algebra_map F E),
+      by { ext, refl }),
+    exact polynomial.splits_comp_of_splits (algebra_map F E) map h_splits },
+  rw [←hα, intermediate_field.adjoin.findim H2],
+  rw ← intermediate_field.card_alg_hom_adjoin_integral F H2 h_separable h_splits,
+  apply fintype.card_congr,
+  transitivity (F⟮α⟯ ≃ₐ[F] F⟮α⟯),
+  { rw hα,
+    change (E ≃ₐ[F] E) ≃ ((⊤ : intermediate_field F E).to_subalgebra ≃ₐ[F]
+      (⊤ : intermediate_field F E).to_subalgebra),
+    rw intermediate_field.top_to_subalgebra,
+    exact
+    { to_fun := λ ϕ, (algebra.top_equiv).trans (trans ϕ (algebra.top_equiv).symm),
+      inv_fun := λ ϕ, (algebra.top_equiv).symm.trans (trans ϕ (algebra.top_equiv)),
+      left_inv := λ _, by { ext, simp only [apply_symm_apply, trans_apply] },
+      right_inv := λ _, by { ext, simp only [symm_apply_apply, trans_apply] } } },
+  { exact alg_equiv_equiv_alg_hom F F⟮α⟯ },
+end
```

Comment:
Done. However, there is an application of `symm_apply_apply` that is painfully slow.

Reply:
On my machine, Lean takes a really long time to finish off the proof, even if the goal is 0 = 0.
https://github.com/leanprover-community/mathlib/pull/4786/commits/bfca5389f25426fa20697d8f1925573cf9cbbe9a

Reply:
In fact, Lean seems to both timeout and say goals accomplished (at the same time)

Reply:
It looks like the line `rw ← intermediate_field.adjoin_simple.card_aut_eq_findim F E H1 h_sep h_splits` causes the timeout, maybe it causes some bad kernel reduction? I'll investigate.

Reply:
Aha: it seems that the kernel is having problems unifying `minimal_polynomial H1` with `minimal_polynomial H2`. If I say `  have h_splits' : (minimal_polynomial H1).splits (algebra_map F F⟮α⟯) := sorry` and use it instead of `h_splits`, there is no timeout. If I see `h_splits'` instead, it timeouts in the kernel.

Reply:
So with the suggestion below, to use `is_galois.is_integral`, `is_galois.separable` and `is_galois.splits`, there should be no more issue here.

Reply:
I could be missing something, but it seems to still be timing out after I made this change.

Reply:
It's very strange. If I include a sorry at the end of the proof then it says goals accomplished, but rightfully complains about the sorry. When I delete the sorry, it times out.

Reply:
The problems have been fixed (after a lengthy discussion in the Berkeley Lean Seminar Zulip stream)